benches/aead.rs

```diff
@@ -0,0 +1,162 @@
+#![allow(missing_docs)]
+
+use criterion::criterion_main;
+use ring::{aead, error};
+
+// All the AEADs we're testing use 96-bit nonces.
+pub const NONCE: [u8; 96 / 8] = [0u8; 96 / 8];
+
+// A TLS 1.2 finished message is always 12 bytes long.
+const TLS12_FINISHED_LEN: usize = 12;
+
+// A TLS 1.3 finished message is "[t]he size of the HMAC output for the
+// Hash used for the handshake," which is usually SHA-256.
+const TLS13_FINISHED_LEN: usize = 32;
+
+// In TLS 1.2, 13 bytes of additional data are used for AEAD cipher suites.
+const TLS12_AD: [u8; 13] = [
+    23, // Type: application_data
+    3, 3, // Version = TLS 1.2.
+    0x12, 0x34, // Length = 0x1234.
+    0, 0, 0, 0, 0, 0, 0, 1, // Record #1
+];
+
+// In TLS 1.3, no additional data is used for AEAD cipher suites.
+const TLS13_AD: [u8; 0] = [];
+
+struct NonceSequence(u64);
+impl NonceSequence {
+    const fn new() -> Self {
+        Self(0)
+    }
+}
+
+impl aead::NonceSequence for NonceSequence {
+    fn advance(&mut self) -> Result<aead::Nonce, error::Unspecified> {
+        let mut result = [0u8; aead::NONCE_LEN];
+        result[4..].copy_from_slice(&u64::to_be_bytes(self.0));
+        self.0 = self.0.checked_add(1).ok_or(error::Unspecified)?;
+        Ok(aead::Nonce::assume_unique_for_key(result))
+    }
+}
+
+macro_rules! bench {
+    ( $benchmark_name:ident, $algorithm:expr, $chunk_len:expr, $ad:expr ) => {
+        pub(super) fn $benchmark_name(c: &mut criterion::Criterion) {
+            use super::{NonceSequence, NONCE};
+            use criterion::{black_box, BatchSize};
+            use ring::{
+                aead::{self, BoundKey},
+                rand::{SecureRandom, SystemRandom},
+            };
+
+            let rng = SystemRandom::new();
+
+            let mut key_bytes = vec![0u8; $algorithm.key_len()];
+            rng.fill(&mut key_bytes).unwrap();
+
+            const FUNC_NAME_BASE: &str =
+                concat!(stringify!($benchmark_name), "_", stringify!($algorithm));
+            let func_name_cleaned = FUNC_NAME_BASE.replace("&aead::", "");
+
+            {
+                let key = aead::UnboundKey::new($algorithm, &key_bytes).unwrap();
+                let mut key = aead::SealingKey::new(key, NonceSequence::new());
+
+                let mut in_out = black_box(vec![0u8; $chunk_len + $algorithm.tag_len()]);
+                let enc_func_name = format!("encrypt_{}", func_name_cleaned);
+                c.bench_function(&enc_func_name, |b| {
+                    b.iter(|| {
+                        let aad = black_box(aead::Aad::from($ad));
+                        key.seal_in_place_append_tag(aad, &mut in_out).unwrap();
```

Comment:
Isn't this going to grow `in_out` longer and longer? I think instead we should use `seal_in_place_separate_tag`.  Then we won't be benchmarking the performance of the memory management of `Vec`.

Maybe it doesn't matter, but I think it would be better to put the `black_box` for `in_out` within this closure like you did for `aad`. Or otherwise we could use a comment about why it is done differently.

We should avoid `unwrap()` since most callers won't call it and we shouldn't measure it.

Probably `let _foo = key.seal_in_place_separate_tag(...)?; Ok(())` or similar instead.

Reply:
Good points here thanks. To address `in_out` growing larger due to the tag, I've moved this benchmark to also use `iter_batched`, which lets you perform data setup. Every run now uses a fresh, pre-allocated vector to ensure the allocator isn't being measured. 

I've also tweaked the usage of `black_box` to be more consistent. Each benchmark has its ciphertext or plaintext setup as part of the benchmark setup and the AEAD AAD setup to obscure the data itself, not the `Aad::from` construction. 

Finally, the unwrapping was removed in favor of a changing the benchmark closure to return a result.